core/src/main/java/com/linecorp/armeria/server/healthcheck/ScheduledHealthChecker.java

```diff
@@ -0,0 +1,99 @@
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.common.util.AbstractListenable;
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.ScheduledFuture;
+
+class ScheduledHealthChecker extends AbstractListenable<HealthChecker>
+        implements ListenableHealthChecker, SafeCloseable {
+    private final AtomicBoolean isHealthy = new AtomicBoolean(false);
+    private final Supplier<? extends CompletionStage<Boolean>> healthChecker;
+    private final EventExecutor eventExecutor;
+    private final Backoff backoff;
+    private final boolean scheduleAfterCheckerComplete;
+    @VisibleForTesting
+    final Set<Future<?>> inScheduledFutures = ConcurrentHashMap.newKeySet();
+
+    ScheduledHealthChecker(Supplier<? extends CompletionStage<Boolean>> healthChecker, Duration period,
+                           double jitter, boolean scheduleAfterCheckerComplete, EventExecutor eventExecutor) {
+        this.healthChecker = healthChecker;
+        this.scheduleAfterCheckerComplete = scheduleAfterCheckerComplete;
+        this.eventExecutor = eventExecutor;
+        backoff = Backoff.fixed(period.toMillis()).withJitter(jitter);
+
+        eventExecutor.execute(createTask());
+    }
+
+    @Override
+    public boolean isHealthy() {
+        return isHealthy.get();
+    }
+
+    @Override
+    public void close() {
+        for (Future<?> future : inScheduledFutures) {
+            future.cancel(true);
+        }
+    }
+
+    private Runnable createTask() {
+        return () -> {
+            if (!scheduleAfterCheckerComplete) {
+                scheduleHealthChecker();
+            }
+            healthChecker.get().handle((result, throwable) -> {
+                final boolean isHealthy;
+                if (throwable != null) {
+                    isHealthy = false;
+                } else {
+                    isHealthy = result;
+                }
+                final boolean oldValue = this.isHealthy.getAndSet(isHealthy);
+                if (oldValue != isHealthy) {
+                    notifyListeners(this);
+                }
+                if (scheduleAfterCheckerComplete) {
+                    scheduleHealthChecker();
+                }
+                return null;
+            });
+        };
+    }
+
+    private void scheduleHealthChecker() {
+        final ScheduledFuture<?> future = eventExecutor.schedule(createTask(), backoff.nextDelayMillis(1),
+                                                                 TimeUnit.MILLISECONDS);
+        inScheduledFutures.add(future);
```

Comment:
How about removing this `inScheduledFutures` and using:
```java
@Nullable
private ScheduledFuture<?> future;
```
instead?
I think we don't need to use a `Set` because there will be one `ScheduledFuture` at most.

Reply:
Originally I used this for fixed rate also, as u comment above, if there is an outlier check, we may has wrong record.